halo2_proofs/examples/vector-mul-parallel-synth.rs

```diff
@@ -0,0 +1,315 @@
+            //
+            // | a0  | a1  | a2  | s_mul |
+            // |-----|-----|-----|-------|
+            // | lhs | rhs | out | s_mul |
+            //
+            let lhs = meta.query_advice(advice[0], Rotation::cur());
+            let rhs = meta.query_advice(advice[1], Rotation::cur());
+            let out = meta.query_advice(advice[2], Rotation::cur());
+            let s_mul = meta.query_selector(s_mul);
+
+            // Finally, we return the polynomial expressions that constrain this gate.
+            // For our multiplication gate, we only need a single polynomial constraint.
+            //
+            // The polynomial expressions returned from `create_gate` will be
+            // constrained by the proving system to equal zero. Our expression
+            // has the following properties:
+            // - When s_mul = 0, any value is allowed in lhs, rhs, and out.
+            // - When s_mul != 0, this constrains lhs * rhs = out.
+            vec![s_mul * (lhs * rhs - out)]
+        });
+
+        FieldConfig {
+            advice,
+            instance,
+            s_mul,
+        }
+    }
+}
+// ANCHOR_END: chip-config
+
+// ANCHOR: chip-impl
+impl<F: Field> Chip<F> for FieldChip<F> {
+    type Config = FieldConfig;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+// ANCHOR_END: chip-impl
+
+// ANCHOR: instructions-impl
+/// A variable representing a number.
+#[derive(Clone, Debug)]
+struct Number<F: Field>(AssignedCell<F, F>);
+
+impl<F: Field> NumericInstructions<F> for FieldChip<F> {
+    type Num = Number<F>;
+
+    fn load_private(
+        &self,
+        mut layouter: impl Layouter<F>,
+        values: &[Value<F>],
+    ) -> Result<Vec<Self::Num>, Error> {
+        let config = self.config();
+
+        layouter.assign_region(
+            || "load private",
+            |mut region| {
+                values
+                    .iter()
+                    .enumerate()
+                    .map(|(i, value)| {
+                        region
+                            .assign_advice(|| "private input", config.advice[0], i, || *value)
+                            .map(Number)
+                    })
+                    .collect()
+            },
+        )
+    }
+
+    fn mul(
+        &self,
+        mut layouter: impl Layouter<F>,
+        a: Vec<Self::Num>,
+        b: Vec<Self::Num>,
+    ) -> Result<Vec<Self::Num>, Error> {
+        let config = self.config();
+        assert_eq!(a.len(), b.len());
+        layouter.assign_region(
+            || "mul",
+            |region: Region<'_, F>| {
+                let thread_safe_region = std::sync::Mutex::new(region);
+                a.par_iter()
+                    .zip(b.par_iter())
+                    .enumerate()
+                    .map(|(i, (a, b))| {
+                        let mut region = thread_safe_region.lock().unwrap();
```

Comment:
I'm a little bit confused here, shouldn't `thread_safe_region` mean that we don't need a lock? IIUC here we don't really paralize the assignment because when a lock is taken by a thread, the others have to wait until it finishes its assignement and continue, which should be slower than normal synthesize.

I tried to run these 2 commands and print out the `synthesize` time
- `cargo run --release --example vector-mul`
- `cargo run --features=thread-safe-region --release --example vector-mul-parallel-synth`

And I found that the latter is indeed slower (roughly 20ms vs 60ms). Is this result expected?

Reply:
Hey @han0110 thank you for pointing this out. The `Mutex` does indeed block full concurrency in a loop and thus the overhead from parallelism means that in loops that _only_ contain assignments we lose out on performance. 
As you've noted -- the benchmarks illustrate this overhead really well. 

Generally this implementation makes `Region` thread-safe in two aspects: firstly it _is_ Send and Sync so can be safely wrapped by a `Mutex`. Prior to these updates Region was not `Send` so couldn't even be wrapped in a Mutex to be inserted into parallel loops. Secondly it makes the copy constraint cycles generated deterministic _even_ when generated in parallel. 

Now passing around mutable references between threads in rust comes with special constraints. For instance `&mut T` is Sync if and only if `&&mut T` is Send (so it's eventually immutable in other threads). The tldr is that if we want to mutate a variable safely the rust compiler will want to enforce exclusive access to it (either by way of a `Mutex` or an `Atomic`, depending on the variable). I think there are other ways to do this but from what I can tell this would require the usage of unsafe code // raw pointers -- which I think we would want to avoid. 

As you point out, if the overhead of parallelism make assigning slower than in the single threaded case, why bother at all. Well there are instances that are more akin to the following: 

```rust 
.zip(b.par_iter())
                    .enumerate()
                    .map(|(i, (a, b))| {
                        ... (a bunch of really expensive operations here)
                    
                        // now lock the region to do assignment
                        let mut region = thread_safe_region.lock().unwrap();

```

basically if the loop has other really expensive calculations and assignment is _not_ the bottleneck you still want to be able to assign in parallel. Our workaround atm is that we 
 - pre-calculate the witness in parallel and hold it in a giant vector  
 - then we serially assign the calculated witness using a regular for loop. 

At the scale we're operating at this is terrible for memory because we need to hold the entire witness in memory -- a much better model is to calculate the witness in parallel AND concurrently assign it using this branch -- whilst discarding the parts of the calculated witness we don't need for the rest of the circuit. This is much more memory efficient and somewhat faster on our end because it only involves a single loop over elements (rather than two). 

Hope that makes sense

Reply:
> basically if the loop has other really expensive calculations and assignment is not the bottleneck you still want to be able to assign in parallel.

I see it now, I didn't think about this kind of case, it make complete sense to me. Thanks for explaining in such detail 🙏 Will do another round of review soon.